src/harness/fakes.ts

```diff
@@ -267,13 +270,11 @@ namespace fakes {
             const document = new documents.TextDocument(fileName, content);
             document.meta.set("fileName", fileName);
             this.vfs.filemeta(fileName).set("document", document);
-            const index = this.outputs.findIndex(output => this.vfs.stringComparer(document.file, output.file) === 0);
-            if (index < 0) {
-                this.outputs.push(document);
-            }
-            else {
-                this.outputs[index] = document;
+            const canonical = ts.toPath(fileName, this.sys.getCurrentDirectory(), ts.createGetCanonicalFileName(this.sys.useCaseSensitiveFileNames));
+            if (!this.outputsMap.has(canonical)) {

```

Comment:
The canonicalization we do is very aggressive, which is one of the reasons `CompilerResult` uses a `SortedMap`. Also, this would completely mask the actual paths we use when we emit files.

Reply:
Not true - the path is only used as a cache key, the document itself still has it's originally assigned path.

Reply:
My apologies, as you are correct on my second point. What I meant to say was that I was thinking of something more like this (which wouldn't regenerate `output` on every dot-property access):

```ts
export class CompilerHost ... {
  ...
  private _outputs: documents.TextDocument[] = [];
  private _outputsIndices: collections.SortedMap<string, number>;

  constructor(...) {
    ...
    this._outputsIndices = new collections.SortedMap<string, number>(sys.vfs.stringComparer);
  }

  public get outputs(): ReadonlyArray<documents.TextDocument> {
    return this._outputs;
  }

  public writeFile(...) {
    ...
    let index = this._outputsIndices.get(document.file);
    if (index === undefined) {
        index = this._outputs.length;
        this._outputs.push(document);
        this._outputsIndices.set(document.file, index);
    }
    else {
        this._outputs[index] = document;
    }    
  }
```

Note that `sys.vfs.stringComparer` actually performs _path_ comparisons based on the case sensitivity of the vfs. As a result, it compares things like `C:\A\B\..\C` and `c:/a/c` as equal on a case-insensitive vfs.